mlir/tools/rocmlir-gen/rocmlir-gen.cpp

```diff
@@ -1429,60 +1535,67 @@ createCPUConvFunc(ModuleOp module,
   return func;
 }
 
-std::string getTypeStr(const mlir::Type &type) {
-  std::string typeName;
-  llvm::raw_string_ostream nameStream(typeName);
-  nameStream << type;
-  return nameStream.str();
-}
-
 static func::FuncOp getMemcpyFuncDecl(ModuleOp &module,
-                                      const mlir::Type &srcElemType,
-                                      const mlir::Type &dstElemType) {
+                                      const mlir::MemRefType srcType,
+                                      const mlir::MemRefType destType) {
   OpBuilder b(module.getContext());
 
-  // memcpy_<srcElemType>_<dstElemType>
+  assert(srcType.getRank() == 1 && "Memcopy takes 1-D sources");
+  assert(destType.getRank() == 1 && "Memcopy takes 1-D destinations");
+
+  mlir::Type srcElemType = srcType.getElementType();
+  mlir::Type dstElemType = destType.getElementType();
+  // memcpy_<srcElemType>_<dstElemType>_(srcSize|any)
   std::string funcName = "_memcpy_";
-  funcName += getTypeStr(srcElemType);
-  funcName += "_";
-  funcName += getTypeStr(dstElemType);
+  llvm::raw_string_ostream funcNameStr(funcName);
+  funcNameStr << srcElemType << "_" << dstElemType << "_";
+
+  int64_t numElements = -1;
+  if (srcType.hasStaticShape())
+    numElements = srcType.getNumElements();
+
+  if (numElements != -1)
+    funcNameStr << numElements;
+  else
+    funcNameStr << "any";
+
+  if ((numElements == -1 && !destType.hasStaticShape()) ||
+      numElements != destType.getNumElements())
+    assert(0 && "Called for an uneven memcpy");
 
   func::FuncOp func = module.lookupSymbol<func::FuncOp>(funcName);
   if (func) // already exists
     return func;
 
-  auto loc = b.getUnknownLoc();
-
-  auto rSrcType = MemRefType::get({-1}, srcElemType);
-  auto rDstType = MemRefType::get({-1}, dstElemType);
+  mlir::Location loc = b.getUnknownLoc();
 
   // clang-format off
-  // func _memcpy_<srcElemType>_<dstElemType> (%arg0 : memref<?xf32>, %arg1 : memref<?xf16, %arg2 : index) {
-  //   scf.for %i0 = %c0 to %arg2 step %c1 {
+  // func _memcpy_<srcElemType>_<dstElemType>_<size> (%arg0 : memref<sizexf32>, %arg1 : memref<sizexf16>) {
+  //   %size = memref.dim %arg0(%c0)
```

Comment:
where does `%c0` come from?

Reply:
And why you want the `size` as part of the function name?

Reply:
So on `-O3`, the size will be inferred from the function name instead of `%size = memref.dim %arg0(%c0)`?
Well, I am trying to learn your "tricks" so I'd rather you keep it.

Reply:
Ok, so, two things of note:
1. As soon as we run `-canonicalize` on the IR (which happens at least once in our pipeline), that `memref.dim` gets replaced by the constant size from the type.
2. (Having gone and looked, it's not clear what optimization level we're using for our host code, it might be -O0 - I'll have to fix that in a future PR)
3. The idea here - which is partly broken by the fact that we stuck a type conversion in here so it might not work, but still - is that compilers are very good at spotting `memcpy()`, _especially_ `memcpy()` with a constant size, and will duly rewrite it to the most convoluted mess of highly-optimized assembly they can manage ... which is a function of the amount of memory being copied. So having separate functions with different constant sizes could give the host code bits a chance to do something very, very, clever

Reply:
I see. Thank you for the explanation.